File: user-service/src/main/java/com/github/manosbatsis/services/user/service/UserServiceImpl.java
package com.github.manosbatsis.services.user.service;

import com.github.manosbatsis.services.user.exception.UserEmailDuplicatedException;
import com.github.manosbatsis.services.user.exception.UserNotFoundException;
import com.github.manosbatsis.services.user.model.User;
import com.github.manosbatsis.services.user.repository.UserRepository;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;

import java.util.List;

@RequiredArgsConstructor
@Service
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;

    @Override
    public List<User> getUsers() {
        return userRepository.findAll();
    }

    @Override
    public User saveUser(User user) {
        return userRepository.save(user);
    }

    @Override
    public void deleteUser(User user) {
        // soft delete, see User annotations
        userRepository.delete(user);
    }

    @Override
    public User validateAndGetUserById(Long id) {
        return userRepository.findById(id).orElseThrow(() -> new UserNotFoundException());
    }

    @Override
    public void validateUserExistsByEmail(String email) {
        userRepository
                .findUserByEmail(email)
                .ifPresent(
                        user -> {
                            throw new UserEmailDuplicatedException(
                                    String.format("User with email '%s' already exist.", email));
                        });
    }
}

File: user-service/src/main/kotlin/com/github/manosbatsis/services/user/service/UserServiceImpl.kt
package com.github.manosbatsis.services.user.service

import com.github.manosbatsis.services.user.exception.UserEmailDuplicatedException
import com.github.manosbatsis.services.user.exception.UserNotFoundException
import com.github.manosbatsis.services.user.model.User
import com.github.manosbatsis.services.user.repository.UserRepository
import org.springframework.stereotype.Service

@Service
class UserServiceImpl(
    private val userRepository: UserRepository,
) : UserService {
    override fun findAll(): List<User> = userRepository.findAll()

    override fun saveUser(user: User): User {
        return userRepository.save(user)
    }

    override fun deleteUser(user: User) {
        // soft delete, see User annotations
        userRepository.delete(user)
    }

    override fun validateAndGetUserById(id: Long): User {
        return userRepository.findById(id).orElseThrow { UserNotFoundException() }
    }

    override fun emailAvailableOrThrow(email: String) {
        userRepository.findUserByEmail(email)?.let {
            throw UserEmailDuplicatedException(
                String.format(
                    "User with email '%s' already exist.",
                    email,
                ),
            )
        }
    }
}
